### PR TITLE
Refactor fusing

### PR DESCRIPTION
**Pull request by reuvenperetz**
## Pull Request Description:
This PR introduces handling graph fusion by encapsulating fusion metadata within a new wrapper class for the graph. This class ensures that after every access to the graph, a validation check is performed to verify that the fusion information remains consistent and that no modifications have introduced inconsistencies. Additionally, the fusion-related logic has been refactored into a new class called `GraphFuser`, which takes the graph along with its fusion metadata and creates a new graph where fused operations are represented as single nodes.

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_facade.py [hunk 1]
```diff
@@ -82,6 +83,11 @@ def search_bit_width(graph_to_search_cfg: Graph,
 
     # Set graph for MP search
     graph = copy.deepcopy(graph_to_search_cfg)  # Copy graph before searching
+    # TODO: The handle of mixed precision with the fused graph will be in a separate PR. Currently, the bit-width
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_facade.py:86]: also needs to be integrated with the BOPs virtual graph

File: model_compression_toolkit/core/common/substitutions/batchnorm_reconstruction.py [hunk 2]
```diff
@@ -143,6 +143,20 @@ def substitute(self,
                                                                     AttributeQuantizationConfig(
                                                                         enable_weights_quantization=False)))
 
+        # Check if the source node was part of a fusion. If so, there are two cases:
+        # either this is no longer a fusion, and the fusion info should be updated by removing
+        # the current info, or this creates a new fusion and the old pattern should be
+        # replaced with the new one.
+        fused_op = graph.get_fusing_info().get_fused_node_name(source_node.name)
+        if fused_op:
+            fused_nodes = graph.get_fusing_info().get_fused_nodes(fused_op)
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/substitutions/batchnorm_reconstruction.py:152]: why the multiple calls for ` graph.get_fusing_info()`?

File: model_compression_toolkit/qat/pytorch/quantization_facade.py [hunk 3]
```diff
@@ -181,6 +183,7 @@ def pytorch_quantization_aware_training_init_experimental(in_model: Module,
 
         user_info.mixed_precision_cfg = bit_widths_config
 
+        qat_model.graph = qat_model.graph.get_internal_graph()
```
Comment (ofirgo) [inline model_compression_toolkit/qat/pytorch/quantization_facade.py:186]: why does the returned model contain the fused graph and not the original?
I'm not sure that I understand what is going on in back2fw and here, and also, this is kind of criptyc (sort of a hidden assumption that can cause problems)

**Review by ofirgo — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/common/fusion/fusing_info.py [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+        for op_id, nodes in self._fusing_data.items():
+            for node in nodes:
+                self._node_to_fused_node_map[node.name] = op_id
+
+    def add_fused_operation(self, op_id: str, nodes: List[BaseNode]) -> None:
+        """
+        Add a new fused operation with the given ID and set of nodes.
+
+        Args:
+            op_id (str): The identifier for the fused operation.
```
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply model_compression_toolkit/core/common/fusion/fusing_info.py:57]: It is what the method generate_fused_op_id generates (the names of the fused nodes concatenated with a prefix of "FusedNode_")

File: model_compression_toolkit/core/common/fusion/fusing_info.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+        return any(node in nodes for nodes in self._fusing_data.values())
+
+    def get_all_fused_operations(self) -> Dict[str, List[BaseNode]]:
+        """
+        Retrieve a copy of all fused operations.
+
+        Returns:
+            Dict[str, Set[BaseNode]]: A copy of the fusing data.
+        """
+        return copy.deepcopy(self._fusing_data)
```
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply model_compression_toolkit/core/common/fusion/fusing_info.py:140]: You correct. I think it is better to return the object without coping it. I'll fix it.

File: model_compression_toolkit/core/common/fusion/fusing_info.py [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+                            )
+                    elif i > 0 and pred not in graph.get_prev_nodes(
+                            nodes[0]):  # External predecessors allowed only for first node
+                        raise ValueError(
+                            f"Fused operation {op_id} contains an external predecessor {pred.name} "
+                            f"to node {node.name}, which is invalid."
+                        )
+
+
+    def get_nodes_to_disable_act_quantization(self):
```
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply model_compression_toolkit/core/common/fusion/fusing_info.py:243]: Currently, it is disabled. I leave the current behavior as is, and in a future PR, this will be removed. There is a TODO for it in the init of the FusingInfo.
As for why it's part of the FusingInfo, I can take it out to a separate function that receives the FusingInfo and returns the nodes that their activation quantization should be disabled.

File: model_compression_toolkit/core/common/fusion/graph_with_fusing_metadata.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+            fusing_info: Dict mapping fused operation IDs to sets of node objects.
+        """
+        assert isinstance(graph, Graph)
+        self._internal_graph = graph
+        self._fusing_info = fusing_info
+        self._fusing_info.validate(graph)  # Ensure initial consistency
+        # TODO: temp disable activation quantization to keep similar functionality. This will be removed in the future
+        self._disable_nodes_activation_quantization()
+
+    # We added __getstate__ and __setstate__ to FusedGraph to fix a recursion error during copy.deepcopy. Without
```
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply model_compression_toolkit/core/common/fusion/graph_with_fusing_metadata.py:51]: It was the original name. I forgot to fix this comment.

File: model_compression_toolkit/core/common/fusion/graph_fuser.py [hunk 8]
```diff
@@ -67,15 +66,29 @@ def _create_fused_node(nodes: List[BaseNode]) -> BaseNode:
...
                               framework_attr={},
                               input_shape=nodes[0].input_shape,
                               output_shape=nodes[-1].output_shape,
-                              weights={},
+                              weights={}, # TODO: update with weights of all nodes
                               layer_class=FusedLayerType)
 
-        # Preserve the final activation quantization configuration
-        # This is important for maintaining the correct behavior of the fused node
+        # Create candidates for this node (we assume that the weights configuration should be taken from the first node, and the activaion configuration
```
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply model_compression_toolkit/core/common/fusion/graph_fuser.py:79]: I'll add assertions.

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_facade.py (hunk 1 above)
Comment (reuvenperetz) [reply model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_facade.py:86]: If I understand correctly, the integration with the BOPs, is simply to build the virtual graph from the fused graph.

File: tests_pytest/_fw_tests_common_base/base_graph_with_fusion_metadata_test.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+        graph_with_fusion_metadata = graph_preparation_runner(self._get_model(),
+                                                              self._data_gen,
+                                                              QuantizationConfig(),
+                                                              fw_info=self.fw_info,
+                                                              fw_impl=self.fw_impl,
+                                                              fqc=self.fqc,
+                                                              mixed_precision_enable=False,
+                                                              running_gptq=False)
+        return graph_with_fusion_metadata
```
Comment (ofirgo) [inline tests_pytest/_fw_tests_common_base/base_graph_with_fusion_metadata_test.py:83]: maybe add verification that the metada exists after running `graph_preparation_runner` and that it is valid or as expected?

File: tests_pytest/_fw_tests_common_base/base_graph_with_fusion_metadata_test.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+
+import model_compression_toolkit.target_platform_capabilities.schema.mct_current_schema as schema
+from model_compression_toolkit.core import QuantizationConfig
+from model_compression_toolkit.core.common import BaseNode
+from model_compression_toolkit.core.common.graph.edge import EDGE_SOURCE_INDEX, EDGE_SINK_INDEX
+from model_compression_toolkit.core.graph_prep_runner import graph_preparation_runner
+
+
+class BaseGraphWithFusingMetadataTest(abc.ABC):
```
Comment (ofirgo) [inline tests_pytest/_fw_tests_common_base/base_graph_with_fusion_metadata_test.py:31]: additional integration test to consider: successful graph update with:
1. with no effect on any fusion in the graph (verify that the fusion validation does not fail us unintentionally)
2. with effect on the graph that is addressed in the substitution (if there such)

**Review by ofirgo — CHANGES_REQUESTED** [review]
consider adding an e2e tests to verify that none of our substitutions fails the fusing metadata.

File: tests_pytest/_fw_tests_common_base/base_graph_with_fusion_metadata_test.py (hunk 10 above)
Comment (reuvenperetz) [reply tests_pytest/_fw_tests_common_base/base_graph_with_fusion_metadata_test.py:31]: Regard 1 - I agree.
Regard 2 - These are the tests below. Or did I misunderstand you?

File: tests_pytest/_fw_tests_common_base/base_graph_with_fusion_metadata_test.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+    fw_impl: Any
```
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply tests_pytest/_fw_tests_common_base/base_graph_with_fusion_metadata_test.py:33]: I used Mixin in the concrete framework classes.

File: tests_pytest/common_tests/unit_tests/core/test_fusion_info.py [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+    assert node_to_fused_map[softmax_node.name] == expected_op2_id
+
+
+def test_fusing_info_validation(mock_graph, fusing_info_generator):
+    """
+    Tests that the fusing info successfully validates a correct graph.
+    - If validation raises an error, the test fails.
+    """
+    fi = fusing_info_generator.generate_fusing_info(mock_graph)
+    fi.validate(mock_graph)
```
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply tests_pytest/common_tests/unit_tests/core/test_fusion_info.py:169]: I'm not sure I understood. Let's talk about it.

File: model_compression_toolkit/core/common/fusion/fusing_info.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply model_compression_toolkit/core/common/fusion/fusing_info.py:140]: I'll change the lists in the info to tuples, and if someone operates on this data, it should copy it from the outside.

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_facade.py [hunk 13]
```diff
@@ -18,6 +18,9 @@
 
 from model_compression_toolkit.core import MixedPrecisionQuantizationConfig
 from model_compression_toolkit.core.common import Graph
+from model_compression_toolkit.core.common.fusion.fusing_metadata_wrapper import FusingMetadataWrapper
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_facade.py:21]: Unnecessary imports?

File: tests/keras_tests/function_tests/test_activation_weights_composition_substitution.py [hunk 14]
```diff
@@ -231,12 +231,12 @@ def test_all_weights_layers_composition(self):
         self.assertTrue(len(sorted_v_nodes[1].candidates_quantization_cfg) == 9)
         # Conv2-Activation node
         self.assertTrue(isinstance(sorted_v_nodes[2], VirtualSplitActivationNode))
-        self.assertTrue(len(sorted_v_nodes[2].candidates_quantization_cfg) == 1)
+        self.assertTrue(len(sorted_v_nodes[2].candidates_quantization_cfg) == 3) # since we do not run fusion, there are three possible activations for the conv
```
Comment (ofirgo) [inline tests/keras_tests/function_tests/test_activation_weights_composition_substitution.py:234]: Doesn't this ruin the test? in there are 3 activations then no composition would happen and the test isn't testing anything.
Why not run fusing?

File: tests_pytest/_test_util/tpc_util.py [hunk 15]
```diff
@@ -50,12 +50,11 @@ def build_mp_config_options_for_kernel_bias_ops(base_w_config: AttributeQuantiza
     return QuantizationConfigOptions(quantization_configurations=mp_configs, base_config=base_op_config)
 
 
-def minimal_tpc():
+def minimal_cfg_options():
```
Comment (ofirgo) [inline tests_pytest/_test_util/tpc_util.py:53]: please publish this new util in the team's chat with a link to the relevant file in the PR

File: tests_pytest/_fw_tests_common_base/fusing/base_fusing_info_generator_test.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+from model_compression_toolkit.core.common.fusion.fusing_info import FusingInfo
+from model_compression_toolkit.core.common.fusion.fusing_metadata_wrapper import FusingMetadataWrapper
+from model_compression_toolkit.core.common.graph.edge import EDGE_SOURCE_INDEX, EDGE_SINK_INDEX
+from model_compression_toolkit.core.graph_prep_runner import graph_preparation_runner
+from tests_pytest._test_util.tpc_util import minimal_cfg_options
+
+
+class BaseFusingInfoGeneratorTest(abc.ABC):
+
+    fw_impl: FrameworkImplementation
```
Comment (ofirgo) [inline tests_pytest/_fw_tests_common_base/fusing/base_fusing_info_generator_test.py:36]: why not use Mixin?
(not sure it is possible, I'm still trying to understand this mechanism)

File: tests_pytest/_fw_tests_common_base/base_graph_with_fusion_metadata_test.py [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+        raise NotImplementedError()
+
+    @pytest.fixture
+    def minimal_tpc_with_fusing(self, default_quant_cfg_options):
+        """
+        Creates a minimal Target Platform Capabilities (TPC) configuration that supports specific fusions.
+        This fixture defines allowed operations and fusing patterns for testing.
+        """
+        return schema.TargetPlatformCapabilities(
+            default_qco=default_quant_cfg_options,
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/_fw_tests_common_base/base_graph_with_fusion_metadata_test.py]: I think that it will be complicated to test this specific attributes on a model in an e2e manner.
in these tests you verify the graph, which is more correct for verifying such edge-cases IMO.
you can leave this extension to the next PR of the e2e tests, and also extend the integration tests then.

File: tests_pytest/common_tests/unit_tests/core/test_fusion_info.py [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+    - Implements `get_next_nodes` and `get_prev_nodes` to maintain linear order.
+    """
+    graph = Mock()
+    graph.nodes.return_value = mock_nodes
+    graph.get_topo_sorted_nodes.return_value = mock_nodes
+
+    def get_next_nodes(node):
+        if node in mock_nodes:
+            index = mock_nodes.index(node)
+            return [mock_nodes[index + 1]] if index < len(mock_nodes) - 1 else []
```
Comment (ofirgo) [inline tests_pytest/common_tests/unit_tests/core/test_fusion_info.py:86]: s it possible do define a more explicit behavior in the mock? it's difficult to follow what is the behavior you expect to get here

**Review by ofirgo — CHANGES_REQUESTED** [review]

File: tests_pytest/_fw_tests_common_base/fusing/base_fusing_info_generator_test.py (hunk 16 above)
Comment (reuvenperetz) [reply tests_pytest/_fw_tests_common_base/fusing/base_fusing_info_generator_test.py:36]: I'm using it. In the concrete class.

File: tests/keras_tests/function_tests/test_activation_weights_composition_substitution.py (hunk 14 above)
Comment (reuvenperetz) [reply tests/keras_tests/function_tests/test_activation_weights_composition_substitution.py:234]: I thought you wanted to remove the fusing from this test, no?

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_facade.py [hunk 19]
```diff
@@ -88,6 +89,11 @@ def search_bit_width(graph: Graph,
     if search_method != BitWidthSearchMethod.INTEGER_PROGRAMMING:
         raise NotImplementedError()
 
+    # Validation is skipped during the mixed-precision search configuration because fusing information is not
+    # relevant for the virtual graph. Therefore, validation checks are disabled before the search begins and
+    # re-enabled once it completes.
+    graph.skip_validation_check = True
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_search_facade.py:95]: maybe replace with a setter function and make this flag a protected member?

**Review by ofirgo — APPROVED** [review]
